pkg/controllers/infracluster/vsphere.go

```diff
@@ -0,0 +1,204 @@
+			Annotations: map[string]string{
+				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+			},
+		},
+		Spec: vspherev1.VSphereClusterSpec{
+			Server: vsphereServerAddr,
+			IdentityRef: &vspherev1.VSphereIdentityReference{
+				Kind: "Secret",
+				Name: r.Infra.Status.InfrastructureName,
+			},
+			ControlPlaneEndpoint: vspherev1.APIEndpoint{
+				Host: apiURL.Hostname(),
+				Port: int32(port),
+			},
+		},
+	}
+
+	if err := r.Create(ctx, target); err != nil {
+		return nil, fmt.Errorf("failed to create InfraCluster: %w", err)
+	}
+
+	log.Info(fmt.Sprintf("InfraCluster '%s/%s' successfully created", defaultCAPINamespace, r.Infra.Status.InfrastructureName))
+
+	return target, nil
+}
+
+// getVSphereMAPIProviderSpec returns a VSphere Machine ProviderSpec from the the cluster.
+func getVSphereMAPIProviderSpec(ctx context.Context, cl client.Client) (*mapiv1beta1.VSphereMachineProviderSpec, error) {
+	rawProviderSpec, err := getRawMAPIProviderSpec(ctx, cl)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain MAPI ProviderSpec: %w", err)
+	}
+
+	providerSpec := &mapiv1beta1.VSphereMachineProviderSpec{}
+	if err := yaml.Unmarshal(rawProviderSpec, providerSpec); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal MAPI ProviderSpec: %w", err)
+	}
+
+	return providerSpec, nil
+}
+
+// ensureVSphereSecret ensures the CAPI VSphere credentials secret exists.
+func (r *InfraClusterController) ensureVSphereSecret(ctx context.Context, vsphereServerAddr string) error {
+	vSphereSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+		},
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(vSphereSecret), vSphereSecret); err != nil && !cerrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get CAPI VSphere credentials secret: %w", err)
+	}
+
+	username, password, err := r.getVSphereCredentials(ctx, vsphereServerAddr)
+	if err != nil {
+		return fmt.Errorf("unable to get VSphere credentials: %w", err)
+	}
+
+	vSphereSecret.StringData = map[string]string{
+		"username": username,
+		"password": password,
+	}
+
+	if err := r.Client.Create(ctx, vSphereSecret); err != nil && !cerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("unable to create CAPI VSphere credentials secret: %w", err)
+	}
+
+	return nil
+}
+
+// getVSphereCredentials obtains the VSphere credentials from the well-known credentials secret.
+func (r *InfraClusterController) getVSphereCredentials(ctx context.Context, vsphereServerAddr string) (string, string, error) {
+	vSphereCredentialsSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: kubeSystemNamespace,
+		Name:      vSphereCredentialsName,
+	}, vSphereCredentialsSecret); err != nil {
+		return "", "", fmt.Errorf("%w %s/%s: %w", errUnableToFindCredsSecret, kubeSystemNamespace, vSphereCredentialsName, err)
```

Comment:
Like I said above, I like the idea of named errors for these cases. However, afaik by wrapping them with `fmt.Errorf`, we can't do any sort of type matching against them later.

Maybe we define a small type for these since main thing that seems to differ is the error message? Something like in https://earthly.dev/blog/golang-errors/.

Reply:
Not sure it is worth doing now, as we are not using those errors anywhere other than where they are thrown from.
I think we can refactor them if we want to export and match on them in the future.

Reply:
When wrapping them with the `%w` verb, you can type match later with existing `errors` package tooling, `errors.As` or `errors.Is`. But, I don't know what it would do with two wraps, lets skip the double wrapping